client/pruning/main.go

```diff
@@ -0,0 +1,118 @@
+	"github.com/cosmos/cosmos-sdk/server"
+	servertypes "github.com/cosmos/cosmos-sdk/server/types"
+	"github.com/cosmos/cosmos-sdk/store/rootmulti"
+	"github.com/tendermint/tendermint/libs/log"
+	dbm "github.com/tendermint/tm-db"
+)
+
+const FlagDBType = "app-db-backend"
+
+// PruningCmd prunes the sdk root multi store history versions based on the pruning options
+// specified by command flags.
+func PruningCmd(appCreator servertypes.AppCreator) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "prune",
+		Short: "Prune app history states based on the pruning options specified by flags",
+		Long: `Prune app history states based on the pruning options specified by flags,
+		Pruning options can be provided via the '--pruning' flag or alternatively with '--pruning-keep-recent', and
+		'--pruning-interval' together.
+		
+		For '--pruning' the options are as follows:
+		
+		default: the last 362880 states are kept, pruning at 10 block intervals
+		nothing: all historic states will be saved, nothing will be deleted (i.e. archiving node)
+		everything: 2 latest states will be kept; pruning at 10 block intervals.
+		custom: allow pruning options to be manually specified through 'pruning-keep-recent', and 'pruning-interval'.
+
+		besides pruning options, database home directory and database backend type should also be specified via flags
+		'--home' and '--app-db-backend'.
+		valid app-db-backend type includes 'goleveldb', 'cleveldb', 'rocksdb', 'boltdb', and 'badgerdb'.
+		`,
+		Example: "prune --home './' --app-db-backend 'goleveldb' --pruning 'custom' --pruning-keep-recent 100 --pruning-interval 10",
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			vp := viper.New()
+
+			// Bind flags to the Context's Viper so we can get pruning options.
+			if err := vp.BindPFlags(cmd.Flags()); err != nil {
+				return err
+			}
+			pruningOptions, err := server.GetPruningOptionsFromFlags(vp)
+			if err != nil {
+				return err
+			}
+
+			home := vp.GetString(flags.FlagHome)
+			db, err := openDB(home, server.GetAppDBBackend(vp))
+			if err != nil {
+				return err
+			}
+
+			logger, _ := log.NewDefaultLogger("plain", "info", false)
+			app := appCreator(logger, db, nil, vp)
+			cms := app.CommitMultiStore()
+
+			if rootMultiStore, ok := cms.(*rootmulti.Store); ok {
+				var pruningHeights []int64
+				latestHeight := rootmulti.GetLatestVersion(db)
+				// valid heights should be greater than 0.
+				if latestHeight <= 0 {
+					logger.Info("the database has no valid heights to prune", "latest_height", latestHeight)
+				}
+
+				shouldBePruned := shouldBePrunedValidator(pruningOptions, latestHeight)
+
+				for height := int64(1); height < latestHeight; height++ {
+					if shouldBePruned(height) {
+						pruningHeights = append(pruningHeights, height)
+						logger.Info("ready to prune", "height", height)
+					}
+				}
+				if len(pruningHeights) == 0 {
+					logger.Info("skip the pruning because no heights")
```

Comment:
```suggestion
					logger.Info("no heights to prune")
					return nil
```

Reply:
Also, we shouldn't be using `logger` for this, but `fmt.Sprintf/Printf`

Reply:
done